swarm_copy_tests/test_agent_routine.py

```diff
@@ -0,0 +1,563 @@
+import json
+from typing import AsyncIterator
+from unittest.mock import patch
+
+import pytest
+from openai.types.chat.chat_completion_chunk import (
+    ChatCompletionChunk,
+    Choice,
+    ChoiceDelta,
+    ChoiceDeltaToolCall,
+    ChoiceDeltaToolCallFunction,
+)
+
+from swarm_copy.agent_routine import AgentsRoutine
+from swarm_copy.new_types import Agent, Response, Result
+from swarm_copy_tests.mock_client import create_mock_response
+
+
+class TestAgentsRoutine:
+    @pytest.mark.asyncio
+    async def test_get_chat_completion(self, mock_openai_client, get_weather_tool):
+        routine = AgentsRoutine(client=mock_openai_client)
+
+        # Test without context variables nor tools
+        agent = Agent()
+        response = await routine.get_chat_completion(
+            agent=agent,
+            history=[{"role": "user", "content": "Hello !"}],
+            context_variables={},
+            model_override=None,
+        )
+        mock_openai_client.assert_create_called_with(
+            **{
+                "model": "gpt-4o-mini",
+                "messages": [
+                    {"role": "system", "content": "You are a helpful agent."},
+                    {"role": "user", "content": "Hello !"},
+                ],
+                "tools": None,
+                "tool_choice": None,
+                "stream": False,
+            }
+        )
+
+        # assert response content
+        assert response.choices[0].message.role == "assistant"
+        assert response.choices[0].message.content == "sample response content"
+
+        # Test with context variables for system prompt but no tools
+        def agent_instruction(context_variables):
+            twng = context_variables.get("twng")
+            mrt = context_variables.get("mrt")
+            return f"This is your new instructions with {twng} and {mrt}."
+
+        agent = Agent(instructions=agent_instruction)
+        response = await routine.get_chat_completion(
+            agent=agent,
+            history=[{"role": "user", "content": "Hello !"}],
+            context_variables={"mrt": "Great mrt", "twng": "Bad twng"},
+            model_override=None,
+        )
+        mock_openai_client.assert_create_called_with(
+            **{
+                "model": "gpt-4o-mini",
+                "messages": [
+                    {
+                        "role": "system",
+                        "content": "This is your new instructions with Bad twng and Great mrt.",
+                    },
+                    {"role": "user", "content": "Hello !"},
+                ],
+                "tools": None,
+                "tool_choice": None,
+                "stream": False,
+            }
+        )
+
+        # assert response content
+        assert response.choices[0].message.role == "assistant"
+        assert response.choices[0].message.content == "sample response content"
+
+        # Test with tools
```

Comment:
I have a general suggestion. I think it is better to separate test cases into different functions. This way you see easier what functionality works and what doesn't. Right now if a big test fails it does not tell you which case works and which case does not.